### PR TITLE
workaround for clang/apple/libc++

### PR DESCRIPTION
**Pull request by DerThorsten**
I see the same issue as mentioned in #108 

This crude workaround/hack allows me to compile on mac with clang

### CODE REVIEW

File: include/sparrow/typed_array.hpp [hunk 1]
```diff
@@ -465,8 +465,25 @@ namespace sparrow
...
+        auto first2 = ta2.begin();
+        for(; first1 != last1; ++first1, ++first2){
+            if (!(*first1 == *first2)){
+                return false;
+            }
+        }
+        return true;
+#else
+        return std::equal(ta1.begin(), ta1.end(), ta2.begin(), ta2.end());
+#endif
```
Comment (SylvainCorlay) [inline include/sparrow/typed_array.hpp:486]: Thanks! I think that the std::equal should work with clang 18. Could we only do the workaround for clang < 18 ?

File: include/sparrow/typed_array.hpp [hunk 2]
```diff
@@ -465,8 +465,25 @@ namespace sparrow
...
-    {
-        return std::equal(ta1.cbegin(), ta1.cend(), ta2.cbegin(), ta2.cend());
+    {   
+        // see https://github.com/man-group/sparrow/issues/108
+#if defined(__APPLE__) && defined(__clang__) && defined(_LIBCPP_VERSION)
+        if(ta1.size() != ta2.size())
+        { 
+            return false;
+        }
+        auto first1 = ta1.begin();
```
Comment (Klaim) [inline include/sparrow/typed_array.hpp:475]: Using `cbegin/cend` would be preferable here.

File: include/sparrow/typed_array.hpp (hunk 1 above)
Comment (SylvainCorlay) [reply include/sparrow/typed_array.hpp:487]: Correct

Comment (DerThorsten) [reply include/sparrow/typed_array.hpp:487]: Any suggestions how to detect libc++<18?

chatgtp suggests some date based versioning shenanigans for LIBCPP
`#if defined(__APPLE__) && defined(__clang__) && defined(_LIBCPP_VERSION) && (_LIBCPP_VERSION < 20220701) `

Comment (jjerphan) [reply include/sparrow/typed_array.hpp:487]: Based on [the definition of `_LIBCPP_VERSION`](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__config#L27-L30), this should work:

```c++
#if defined(__APPLE__) && defined(__clang__) && defined(_LIBCPP_VERSION) && (_LIBCPP_VERSION < 190000)
```

Comment (Klaim) [reply include/sparrow/typed_array.hpp:486]: My understanding is that `defined(__APPLE__) && defined(__clang__) && ` can then be removed. `libc++` can be used by other compilers than clang (like GCC and mingw, not sure about msvc) and (IMUIC) this is not specific to Apple platforms.

**Review by JohanMabille — APPROVED** [review]

**Review by Klaim — APPROVED** [review]
LGTM 👍🏽 in particular if it works with https://github.com/man-group/sparrow/pull/114

**Review by SylvainCorlay — APPROVED** [review]

**Pull request merged**